src/nvim/edit.c

```diff
@@ -3672,28 +3672,28 @@ static buf_T *ins_compl_next_buf(buf_T *buf, int flag)
 }
 
 
-/*
- * Execute user defined complete function 'completefunc' or 'omnifunc', and
- * get matches in "matches".
- */
+// Execute user defined complete function 'completefunc' or 'omnifunc', and
+// get matches in "matches".
 static void
-expand_by_function (
-    int type,                   /* CTRL_X_OMNI or CTRL_X_FUNCTION */
+expand_by_function(
+    int type,                   // CTRL_X_OMNI or CTRL_X_FUNCTION
     char_u *base
 )
 {
-  list_T      *matchlist = NULL;
-  dict_T      *matchdict = NULL;
-  char_u      *funcname;
+  list_T *matchlist = NULL;
+  dict_T *matchdict = NULL;
+  char_u *funcname;
   pos_T pos;
-  win_T       *curwin_save;
-  buf_T       *curbuf_save;
+  win_T *curwin_save;
+  buf_T *curbuf_save;
   typval_T rettv;
   const int save_State = State;
 
+  assert(curbuf != NULL && curbuf->b_p_cfu != NULL && curbuf->b_p_ofu != NULL);
   funcname = (type == CTRL_X_FUNCTION) ? curbuf->b_p_cfu : curbuf->b_p_ofu;
-  if (*funcname == NUL)
+  if (*funcname == NUL) {
     return;
```

Comment:
JFI: this code (`return`) appears to never be used (in tests).

Reply:
Which makes sense according to the above `assert`..
So this appears to be dead code now?

Reply:
Not sure if it has to do with assert..
`NUL != NULL`

Reply:
yeah the ternary would be dead code if we are assetings `curbuf->b_p_cfu != NULL` just before it. 

The clang warning wasn't clear to me, I thought it was only warning about `curbuf` being NULL. So let's only assert  that, and see if clang still warns after that.

    assert(curbuf != NULL);

Reply:
fixed

Reply:
What I've (initially) meant is that the `*funcname == NUL` case is not covered (in tests).
Out of scope for this PR, but just wanted to note it.

Reply:
@justinmk both `b_p_cfu` and `b_p_ofu` are also pointers, which may create another clang warning at `*funcname`

Reply:
@ngortheone then maybe the existing if-check should check `if (funcname == NULL || *funcname == NUL)`

Reply:
..and should also be considered to be forwarded / be done via Vim then maybe.
(including having a test for the missing case, if that's also not covered there - filing an issue for it might enough to get the patch for it.  We might just miss a test for this though (or it is tested via disabled/skipped patch))